---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# Differential Binding (dbr)
## Shanahan Group at UZH and Levitt TFBM data


<!-- [![Travis build status](https://travis-ci.org/chumbleycode/dbr.svg?branch=master)](https://travis-ci.org/chumbleycode/dbr) -->

[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/chumbleycode/dbr?branch=master&svg=true)](https://ci.appveyor.com/project/chumbleycode/dbr)

The goal of dbr is to implicate some gene regulator as the upstream cause of differential RNA expression between differerent treatment groups. Colloquially, whether there is "differential binding" (DB) of the regulator over treatments. In practice, dbr asks whether the pattern of differential RNA expression over genes reflects (the binding-site availability of) some upstream regulator. Binding-site availability or "enrichment" just refers to the total count of binding sites found in the DNA of each gene, which is in turn derived from external genomic DNA. 

The package is in development and is likely to change. It currently reimplements the important TeLiS method of Cole et. al. (2005), which used a gene-set approach to infer DB. Our reimplementation incorporates the most up-to-date motif binding data. Additionally, our package offers a non-parametric version of TeLiS, and provides simple new functionality that eschews the need to heuristically define a gene set altogether, i.e. a set that is categorically "differentially expressed", before proceeding to DB analysis. Instead we simply regress gene-specific DE estimates on gene-specific binding-site counts over the entire relevant genome (the set of genes with at least one binding motif for at least one regulator). For computational convenience our regression approach currently avoids full multilevel modeling. 

## Installation

You can install dbr with:

```r
install.packages("devtools")
devtools::install_github("chumbleycode/dbr")
library(dbr)
```

## Data: a simple example

There are currently three TFBM matrices: utr1, exonic1, exonic_utr1. Get more info for each via ?utr1, ?exonic1, etc.

Motifs of interest may be found in the columns of these matrices. For example, recent literature has examined "a pre-specified set of TFs involved in inflammation (NF-kB and AP-1), IFN response (interferon-stimulated response elements; ISRE), SNS activity (CREB, which mediates SNS-induced b-adrenergic signaling), and glucocorticoid signaling (glucocorticoid receptor; GR)."  In biomart nomenclature, "NF-kB" is is identified with NFKB1 or NFKB2.  AP-1 is called JUN. ISRE is identified with set of tfs including IRF2, IRF3, IRF4, 5, 7, 8, 9. CREB identified with CREB3 or CREB3L1. GR is called NR3C1. This leaves us with 13 regulators plus one complex CEBPG::CREB3L1 (CEBPG_CREB3L1), as follows.

```r
immune_tfbms = c("CEBPG_CREB3L1", "CREB3", "CREB3L1", "IRF2", "IRF3", "IRF4", "IRF5", "IRF7", "IRF8", "IRF9", "JUN", "NFKB1", "NFKB2", "NR3C1")
utr1[, immune_tfbms] # the gene-by-motif matrix for immune motifs of interest
```

## Analysis: a simple example.

We examine DB of some immune regulators amoung people with early-life stress (relative to unstressed) using data from Cole et al. (2016). Such analyses generally have two steps. 

1. Differential expression (DE): Estimate differential RNA expression across exposure groups. Here we use a linear model: the exposure must be a *single* column of "design" matrix X (dbr cannot currently handle treatments defined across multiple collumns, e.g. factors with many levels). 
2. Differential binding (DB): Infer dependence of this estimate (over genes) on the binding-site count.

#### DE

```r
# Load packages
library(tidyverse)
library(limma)

# Download open source data then specify gene-by-gene regression model
dat = GEOquery::getGEO("GSE77164")[[1]]

# Specify whole-genome regression of rna on design
y <- dat %>% Biobase::exprs()
X <- dat %>%
Biobase::pData() %>%
select(age = `age:ch1`,
soldier = `childsoldier:ch1`,
edu = `educationlevel:ch1`)
X     <- model.matrix(~ soldier + edu + age, data = X) 

# Estimate DE using standard limmma/edger pipeline. 
of_in <- "soldier1"
ttT <-
lmFit(y, X) %>%
eBayes %>%
tidy_topTable(of_in = of_in)
```

#### DB

Perhaps the simplest differential binding analysis is just a regression of gene-wise DE estimates on motif site count 

##### Regression approach


```r
# Note the number of genes in the sample frame, the number containing at least 1 DNA motif, and the intersection of these gene-sets.
length(ttT$gene)
length(rownames(utr1))
length(intersect(rownames(utr1), ttT$gene))

# Inspect the tfbm-augmented DE results: note the dimensions
ttT %>% append_db(which_matrix = utr1) 
# To include explicit zeros
ttT %>% append_db(which_matrix = utr1, explicit_zeros = T)

# regress DE on one motif of interest
summary(lm(B ~ NR3C1, data = append_db(ttT))) 
summary(lm(B ~ NR3C1, data = append_db(ttT, explicit_zeros = T))) # note: different degrees of freedom
summary(lm(B ~ NR3C1, data = append_db(ttT, which_matrix = exonic1))) # note: different binding location 

# REGRESS DE ESTIMATE ON MOTIF SITE COUNT FOR THE TRANSCRIPTION FACTOR: BEWARE MULTIPLICITY
ttT %>%
infer_db(which_tfbms = immune_tfbms) %>%
extract_db

# EXPLORATORY FISHING: BEWARE MULTIPLICITY
ttT %>%
infer_db %>%
extract_db 

```

##### An alternative approach (see Cole et al)

This approach requires some heuristic to select a set of subgenes to categorically label as DE. For example, we can filter on uncorrected p.value, or logfc (not an inference). This subset is identified by the argument ttT_sub to infer_db(). ttT_sub is just a scrict subset of the rows of our universe of genes, given in ttT above (e.g. the subset whose logFC, or uncorrect p value exceeds some heuristic value). The additional options blow indicate that we have select our prefered tfbm matrix "which_matrix", tfbm hypothesis set "which_tfbms", and methods (p_npar, p_par, which require tt_sub be specified).

```r
ttT %>%
infer_db(ttT_sub        = filter(ttT, P.Value <= 0.05),
which_matrix   = exonic1_utr1 ,
which_tfbms    = immune_tfbms,
explicit_zeros = TRUE) %>%
extract_db(methods =  c("p_npar", "p_par")) 
``` 
